include/engine/routing_algorithms/routing_base.hpp

```diff
@@ -573,29 +634,46 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
         BOOST_ASSERT_MSG((SPECIAL_NODEID != middle && INVALID_EDGE_WEIGHT != distance),
                          "no path found");
 
-        // we need to unpack sub path from core heaps
-        if (facade->IsCoreNode(middle))
+        if (distance != forward_heap.GetKey(middle) + reverse_heap.GetKey(middle))
         {
-            std::vector<NodeID> packed_core_leg;
-            RetrievePackedPathFromHeap(forward_core_heap, reverse_core_heap, middle,
-                                       packed_core_leg);
-            BOOST_ASSERT(packed_core_leg.size() > 0);
-            RetrievePackedPathFromSingleHeap(forward_heap, packed_core_leg.front(), packed_leg);
-            std::reverse(packed_leg.begin(), packed_leg.end());
-            packed_leg.insert(packed_leg.end(), packed_core_leg.begin(), packed_core_leg.end());
-            RetrievePackedPathFromSingleHeap(reverse_heap, packed_core_leg.back(), packed_leg);
+            // self loop
+            BOOST_ASSERT(forward_heap.GetKey(middle) + reverse_heap.GetKey(middle) < 0);
+            BOOST_ASSERT(forward_heap.GetData(middle).parent == middle &&
+                         reverse_heap.GetData(middle).parent == middle);
+            packed_leg.push_back(middle);
+            packed_leg.push_back(middle);
         }
         else
         {
-            RetrievePackedPathFromHeap(forward_heap, reverse_heap, middle, packed_leg);
+            // we need to unpack sub path from core heaps
+            if (facade->IsCoreNode(middle))
+            {
+                std::vector<NodeID> packed_core_leg;
+                RetrievePackedPathFromHeap(forward_core_heap, reverse_core_heap, middle,
+                                           packed_core_leg);
+                BOOST_ASSERT(packed_core_leg.size() > 0);
+                RetrievePackedPathFromSingleHeap(forward_heap, packed_core_leg.front(), packed_leg);
+                std::reverse(packed_leg.begin(), packed_leg.end());
```

Comment:
needs algorithm header; looking at the file, several headers are missing, e.g. vector, iterator, utility, limits, cstddef